homeassistant/components/roborock/roborock_storage.py

```diff
@@ -0,0 +1,118 @@
+"""Roborock storage."""
+
+import dataclasses
+import logging
+import os
+import time
+
+from homeassistant.core import HomeAssistant
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+MAP_PATH = f"{DOMAIN}/maps"
+MAP_UPDATE_FREQUENCY = 3600  # Only save the map once every hour.
+
+
+async def get_roborock_storage(hass: HomeAssistant, entry_id: str):
+    """Get a roborock storage object for a given config entry."""
+    map_path = hass.config.path(MAP_PATH)
+
+    return RoborockStorage(hass, map_path, entry_id)
+
+
+@dataclasses.dataclass
+class RoborockMapEntry:
+    """Describe a map entry stored on disk."""
+
+    map_name: str
+    content: bytes
+    time: float
+
+
+class RoborockStorage:
+    """Store Roborock data."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        map_path: str,
+        entry_id: str,
+    ) -> None:
+        """Initialize RoborockStorage."""
+        self._hass = hass
+        self._data: dict[str, RoborockMapEntry] = {}
+        self._map_path = map_path
+        self._entry_id = entry_id
+
+    def _should_update(self, map_entry: RoborockMapEntry | None, content: bytes):
+        return (
+            map_entry is None
+            or time.time() - map_entry.time > MAP_UPDATE_FREQUENCY
+            or content != map_entry.content
+        )
+
+    def _get_map_filename(self, map_name: str):
+        return self._hass.config.path(f"{MAP_PATH}/{self._entry_id}/{map_name}")
+
+    async def async_load_map(self, map_name: str) -> bytes | None:
```

Comment:
You should create an `async_load_maps` function so we only have a single executor job instead of 3. That will be a lot faster and during startup there are not enough executor threads available